File: sdk/go/http_sender.go
package omniswitch

import (
	"crypto/rand"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

// HttpResult is the normalized result of an HTTP call — status code plus
// raw response body, regardless of whether the status was 2xx or not
// (unlike Go's own http.Client, which doesn't treat a non-2xx status as
// an error at all — this type exists so Client's own status-code
// branching, e.g. the 401-retry-once path, has one shape to work with).
type HttpResult struct {
	StatusCode int
	Body       string
}

// HttpSender is the one seam Client depends on for making HTTP calls —
// injectable so tests can supply a mock without a real network call, the
// same role `fetch` plays in the Node SDK. The default implementation
// (NewHTTPClientSender) uses net/http.
type HttpSender interface {
	Send(method, url string, headers map[string]string, body *string, timeoutMs int) (HttpResult, error)
}

type httpClientSender struct {
	client *http.Client
}

// No external HTTP dependency.
func NewHTTPClientSender() HttpSender {
	return &httpClientSender{client: &http.Client{}}
}

func (s *httpClientSender) Send(method, url string, headers map[string]string, body *string, timeoutMs int) (HttpResult, error) {
	var reader io.Reader
	if body != nil {
		reader = strings.NewReader(*body)
	}

	req, err := http.NewRequest(method, url, reader)
	if err != nil {
		return HttpResult{}, err
	}
	for key, value := range headers {
		req.Header.Set(key, value)
	}

	client := s.client
	client.Timeout = time.Duration(timeoutMs) * time.Millisecond

	resp, err := client.Do(req)
	if err != nil {
		return HttpResult{}, err
	}
	defer resp.Body.Close()

	responseBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return HttpResult{}, err
	}

	return HttpResult{StatusCode: resp.StatusCode, Body: string(responseBody)}, nil
}

// No external dependency for something this small.
func newUUIDv4() string {
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		panic(err) // crypto/rand failing is not a recoverable condition
	}
	b[6] = (b[6] & 0x0f) | 0x40 // version 4
	b[8] = (b[8] & 0x3f) | 0x80 // variant 10
	return fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:16])
}

File: sdk/rust/src/http_sender.rs
//! The one seam `OmniSwitchClient` depends on for making HTTP calls —
//! injectable so tests can supply a mock without a real network call,
//! the same role `fetch` plays in the Node SDK.

use std::collections::HashMap;
use std::error::Error;
use std::time::Duration;

#[derive(Debug, Clone)]
pub struct HttpResult {
    pub status_code: u16,
    pub body: String,
}

pub trait HttpSender: Send + Sync {
    fn send(
        &self,
        method: &str,
        url: &str,
        headers: &HashMap<String, String>,
        body: Option<&str>,
        timeout_ms: u64,
    ) -> Result<HttpResult, Box<dyn Error>>;
}

/// Default [`HttpSender`] — `ureq` (synchronous, minimal dependency
/// surface — no async runtime forced on callers), not tied to any one
/// async executor.
pub struct UreqHttpSender;

impl HttpSender for UreqHttpSender {
    fn send(
        &self,
        method: &str,
        url: &str,
        headers: &HashMap<String, String>,
        body: Option<&str>,
        timeout_ms: u64,
    ) -> Result<HttpResult, Box<dyn Error>> {
        let agent = ureq::AgentBuilder::new().timeout(Duration::from_millis(timeout_ms)).build();
        let mut request = agent.request(method, url);
        for (key, value) in headers {
            request = request.set(key, value);
        }

        let result = match body {
            Some(b) => request.send_string(b),
            None => request.call(),
        };

        match result {
            Ok(response) => {
                let status_code = response.status();
                let body_text = response.into_string()?;
                Ok(HttpResult { status_code, body: body_text })
            }
            Err(ureq::Error::Status(status_code, response)) => {
                // ureq treats any non-2xx as an Err — normalize back to a
                // plain HttpResult so the client's own status-code
                // branching (401 retry, non-2xx -> OmniSwitchApiError)
                // doesn't need two different code paths.
                let body_text = response.into_string().unwrap_or_default();
                Ok(HttpResult { status_code, body: body_text })
            }
            Err(e) => Err(Box::new(e)),
        }
    }
}
